### PR TITLE
Add the capability to assimilate the MADIS snow depth data from GTS

### PR DESCRIPTION
**Pull request by jiaruidong2017**
This PR updates the `land_analysis.py` to add the capability to assimilate the madis snow depth data from GTS. 

This PR depends on GDASApp [#602](https://github.com/NOAA-EMC/GDASApp/pull/602)

# How has this been tested?
Clone and build on Hera
Cycled test on Hera Run C384/C192 L127 for 2023050100.
Observations on hera at `/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/para_gfs/glopara_dump/gdas.20230501/`

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: parm/config/gfs/config.preplandobs [hunk 1]
```diff
@@ -10,7 +10,8 @@ echo "BEGIN: config.preplandobs"
 
 export GTS_OBS_LIST="${HOMEgfs}/sorc/gdas.cd/parm/land/prep/prep_gts.yaml"
 export BUFR2IODAX="${HOMEgfs}/exec/bufr2ioda.x"
-export BUFR2IODAYAML="${HOMEgfs}/sorc/gdas.cd/test/testinput/bufr_adpsfc_snow.yaml"
+export BUFRADPSFCYAML="${HOMEgfs}/sorc/gdas.cd/test/testinput/bufr_adpsfc_snow.yaml"
+export BUFRSNOCVRYAML="${HOMEgfs}/sorc/gdas.cd/test/testinput/bufr_snocvr.yaml"
```
Comment (aerorahul) [inline parm/config/gfs/config.preplandobs]: Should these yamls be in `parm/land/` and not `test/`?

File: ush/python/pygfs/task/land_analysis.py [hunk 2]
```diff
@@ -105,14 +110,34 @@ def prepare_GTS(self) -> None:
...
+            rm_p(output_file)
+
+        # execute BUFR2IODAX to convert snocvr bufr data into IODA format
+        yaml_file = f"bufr_snocvr.yaml"
+        if not os.path.isfile(f"{os.path.join(localconf.DATA, yaml_file)}"):
+            logger.exception(f"{yaml_file} not found")
+            raise FileNotFoundError(f"{os.path.join(localconf.DATA, yaml_file)}")
+        exe = Executable(self.task_config.BUFR2IODAX)
+        exe.add_default_arg(os.path.join(localconf.DATA, f"{yaml_file}"))
+
```
Comment (aerorahul) [inline ush/python/pygfs/task/land_analysis.py]: I have a suggestion to remove some repetition here. Please see the patch on this file if you could kindly test.
[diff.txt](https://github.com/NOAA-EMC/global-workflow/files/12563299/diff.txt)

File: ush/python/pygfs/task/land_analysis.py [hunk 3]
```diff
@@ -105,14 +110,34 @@ def prepare_GTS(self) -> None:
...
+        try:
+            exe()
+        except OSError:
+            raise OSError(f"Failed to execute {exe}")
+        except Exception:
+            raise WorkflowException(f"An error occured during execution of {exe}")
+
+        output_file = f"{localconf.OPREFIX}snocvr_snow.nc4"
+        if os.path.isfile(f"{os.path.join(localconf.DATA, output_file)}"):
+            rm_p(output_file)
```
Comment (WalterKolczynski-NOAA) [inline ush/python/pygfs/task/land_analysis.py]:
```suggestion
        rm_p(output_file)
```

**Review by WalterKolczynski-NOAA — DISMISSED** [review]
Each of these different steps should probably be broken into individual methods that `prepare_GTS` calls, as it is getting a bit long now and doing multiple things. Will see what @aerorahul thinks.

File: ush/python/pygfs/task/land_analysis.py [hunk 4]
```diff
@@ -87,12 +87,17 @@ def prepare_GTS(self) -> None:
...
+        adpsfc_yaml = os.path.join(self.runtime_config.DATA, "bufr_adpsfc_snow.yaml")
+        logger.info(f"Generate BUFR2IODA YAML file: {adpsfc_yaml}")
+        temp_yaml = parse_j2yaml(self.task_config.BUFRADPSFCYAML, self.task_config)
+        save_as_yaml(temp_yaml, adpsfc_yaml)
+        logger.info(f"Wrote bufr2ioda YAML to: {adpsfc_yaml}")
+        snocvr_yaml = os.path.join(self.runtime_config.DATA, "bufr_snocvr.yaml")
+        logger.info(f"Generate BUFR2IODA YAML file: {snocvr_yaml}")
+        temp_yaml = parse_j2yaml(self.task_config.BUFRSNOCVRYAML, self.task_config)
+        save_as_yaml(temp_yaml, snocvr_yaml)
+        logger.info(f"Wrote bufr2ioda YAML to: {snocvr_yaml}")
```
Comment (aerorahul) [inline ush/python/pygfs/task/land_analysis.py]: I think we may be able to make this a bit less hard-coded and yaml driven from the top.
Say for e.g. somewhere in GDASApp/parm/land has `bufr2ioda.yaml` that contains:
```yaml
bufr2ioda:
     adpsfc: {{ HOMEgfs }}/sorc/gdas.cd/test/testinput/bufr_adpsfc_snow.yaml
     snocvr: {{ HOMEgfs }}/sorc/gdas.cd/test/testinput/bufr_snocvr.yaml
```
We could then loop over this list to parse the yamls, and call the `bufr2ioda.x` and verify the output is produced.

The code then does not know what it is processing and bufr files can be added/removed to the yaml.

Would this be acceptable?

**Review by aerorahul — COMMENTED** [review]
Mostly looks good.  We can work to make it better and data driven.

File: ush/python/pygfs/task/land_analysis.py (hunk 4 above)
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py]: we may be able to use `land/prep/prep_gts.yaml` as it already seems to have the appropriate information and it is likely better to keep them together.

File: parm/config/gfs/config.preplandobs (hunk 1 above)
Comment (jiaruidong2017) [reply parm/config/gfs/config.preplandobs]: All the GTS bufr2ioda yamls in GDASApp are saved at the `test/testinput/` directory.

File: ush/python/pygfs/task/land_analysis.py (hunk 2 above)
Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: I just tested this, and it worked perfect. The updates were committed to the repo. Thanks @aerorahul for your suggestions.

File: ush/python/pygfs/task/land_analysis.py (hunk 4 above)
Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: @aerorahul Could you please give more clue how to do this? Thanks.

Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: I mean how to loop over this list to parse the yamls?

Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: @aerorahul Follow your suggestions, I have made the changes for your review. Thanks.

File: ush/python/pygfs/task/land_analysis.py (hunk 3 above)
Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: The above lines were removed.

File: ush/python/pygfs/task/land_analysis.py [hunk 5]
```diff
@@ -87,12 +87,13 @@ def prepare_GTS(self) -> None:
...
-        temp_yaml = parse_j2yaml(self.task_config.BUFR2IODAYAML, self.task_config)
-        save_as_yaml(temp_yaml, bufr2ioda_yaml)
-        logger.info(f"Wrote bufr2ioda YAML to: {bufr2ioda_yaml}")
+        # generate bufr2ioda YAML files
+        for name in ["adpsfc", "snocvr"]:
+            gts_yaml = os.path.join(self.runtime_config.DATA, f"bufr_{name}_snow.yaml")
+            logger.info(f"Generate BUFR2IODA YAML file: {gts_yaml}")
+            temp_yaml = parse_j2yaml(prep_gts_config.bufr2ioda[name], localconf)
+            save_as_yaml(temp_yaml, gts_yaml)
+            logger.info(f"Wrote bufr2ioda YAML to: {gts_yaml}")
```
Comment (aerorahul) [inline ush/python/pygfs/task/land_analysis.py]: Thanks for updating this section @jiaruidong2017 
Here is what I was suggesting.  
In the attached patch, you will notice that I have removed any knowledge of what is being converted from  bufr to ioda in the python script.  The control comes directly from the yaml.

Would you be willing to give this a try?
[diff.txt](https://github.com/NOAA-EMC/global-workflow/files/12580243/diff.txt)

Thanks

Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: Thanks @aerorahul. I will try it and let you know then.

File: ush/python/pygfs/task/land_analysis.py [hunk 6]
```diff
@@ -118,11 +110,18 @@ def _gtsbufr2iodax(exe, yaml_file):
...
 
-        # execute BUFR2IODAX to convert snocvr bufr data into IODA format
-        _gtsbufr2iodax(exe, os.path.join(localconf.DATA, "bufr_snocvr_snow.yaml"))
+            # execute BUFR2IODAX to convert {name} bufr data into IODA format
+            _gtsbufr2iodax(exe, os.path.join(localconf.DATA, f"bufr_{name}_snow.yaml"))
```
Comment (aerorahul) [inline ush/python/pygfs/task/land_analysis.py]:
```suggestion
            _gtsbufr2iodax(exe, gts_yaml)
```
In my second (revised diff.txt), I removed the need to construct this again.

File: ush/python/pygfs/task/land_analysis.py (hunk 5 above)
Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: @aerorahul I just tested and it worked perfectly. The changes have been committed to the repo.

File: ush/python/pygfs/task/land_analysis.py (hunk 6 above)
Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: Done the changes.

**Review by aerorahul — APPROVED** [review]
Looks good.
Thank you for accepting suggestions.
If I could make one more suggestion that aids in knowing the inputs of this method completely, is to consolidate everything that is needed into `localconf` and then only use `localconf` everywhere.  
At present, you will notice, variables are being accessed from `task_config`, `runtime_config` as well as `localconf`.

**Comment by jiaruidong2017** [comment]
> Looks good. Thank you for accepting suggestions. If I could make one more suggestion that aids in knowing the inputs of this method completely, is to consolidate everything that is needed into `localconf` and then only use `localconf` everywhere. At present, you will notice, variables are being accessed from `task_config`, `runtime_config` as well as `localconf`.

Thanks @aerorahul for the detailed instructions, and I have benefited a lot when I was working with you on this PR.

**Comment by jiaruidong2017** [comment]
@aerorahul This PR depends on the https://github.com/NOAA-EMC/GDASApp/pull/602. Would you please take time to review it? Thanks again.

**Comment by aerorahul** [comment]
@jiaruidong2017 
Let us know when you want us to merge this PR.  It is good to go from this side.

**Comment by jiaruidong2017** [comment]
@aerorahul we need to wait for @CoryMartin-NOAA to merge the recent updates in GDASApp into the feature/stable-nightly, and update the commit hash in the global-workflow. I will let you know when it is ready. Thanks.

**Comment by CoryMartin-NOAA** [comment]
@jiaruidong2017 already done... [d347d22](https://github.com/NOAA-EMC/GDASApp/commit/d347d22fb91ccdaf947d2d8cddb0e2281e3997d4)

**Comment by jiaruidong2017** [comment]
Thanks @CoryMartin-NOAA I have updated the commit hash for the GDASApp. 
@aerorahul @WalterKolczynski-NOAA It is ready to merge this PR. Thanks.

**Pull request merged**